### ISSUE TITLE
leak detection overwrites test error (OR: subsequent errors can clobber prior in report?)

### ISSUE THREAD

**Issue by ScottFreeCode** [mochajs/mocha#2971]
Run this test file with `--check-leaks`:
```js
it("leak issue", function() {
  leaky = true
  throw new Error("And also fail the test")
})
```

You should get one error "And also fail the test" and another "Error: global leak detected: leaky" (or perhaps, if not both, then only one or the other). Instead you get two of the latter.

I'd guess that the error property on the failed test is being overwritten and an extra fail event emitted for that test, instead of generating a fail event associated with the test but with its own error property.

**Comment by ScottFreeCode** [comment]
This is *probably* just a specific case of a general issue with errors overwriting prior errors, see e.g. the way "second" is reported twice instead of "first" and "second" in #2906.

**Comment by juergba** [comment]
This issue was fixed with #4033.
```
  1) leak issue
  2) leak issue

  0 passing (26ms)
  2 failing

  1) leak issue:
     Error: And also fail the test
      at Context.<anonymous> (D:\mocha-fork\errorCode.js:72:9)
      [...]      

  2) leak issue:
     Error: global leak detected: 'leaky'
      at Runner.checkGlobals (lib\runner.js:273:17)
      [...]
```
